Adding a way to unstuck the game script if needed

Harest opened the pull request:
- Check made every minute to see if the game script is running. If no round started in the last 15 minutes, we try to start a new round.  
- We "deny" any call to update_grid() if a successful one happened in the last 13 seconds.  
- Temporal quick fix for the issue #59.  
- Improvement in the planet switch : if the best planet found has an higher difficulty while there's errors, since we've the data on our current difficulty, we switch.  
- Auto planet switching turned on by default. You can still disable it if you want. But it seems CORS issues are a thing from the past maybe ? This would close issue #22.  
_Edit : So i'll say this fix #59, fix #50 and fix #62._

Harest committed:
Adding a way to unstuck the game script if needed

- Check made every minute to see if the game script is running. If no round started in the last 15 minutes, we try to start a new round.  
- We "deny" any call to update_grid() if a successful one happened in the last 13 seconds.  
- Temporal quick fix for the issue #59.  
- Improvement in the planet switch : if the best planet found has an higher difficulty while there's errors, since we've the data on our current difficulty, we switch.  
- Auto planet switching turned on by default. You can still disabled it if you want. But it seems CORS issues are a thing from the past maybe ? This would close issue #22.

Harest commented:
I forgot something. Maybe we should warn that turning off animations will disable the auto planet switching ? And set `auto_switch_planet.active = false;` when you confirm the disabling of anim. Since planet switching is made with the game logic and it'd cause issues(?).  
Or still set auto planet switching to false by default, even if i think most people using the script would probably prefer the most automatized choice.  
I don't know how this works on the game logic. If it only affects display or not.

ensingm2 committed:
DIsable auto-planet-switch on disable of anims

ensingm2 commented:
Gonna try running this PR for a bit and see if I can break it. Otherwise, merging in.

ensingm2 commented:
My concern with this is snowballing requests. Say we infinite loop again, sending requests every 5s that error out. The check then runs 15 minutes later, and doesn't clear that original interval, so starts another zone attempt concurrently. Since the other one is still sending requests, this one will fail too. 15 minutes later, another will start, etc.

If you leave that running all night erroring out, you could build up to be hitting the servers 5+ times per second. Is that correct thinking?

Harest commented:
Sure. This is what i'm using for quite a while now. Except for the empty JSON responses when submitting a score and the retries that never seem to work (always the same empty JSON response with `x-eresult | 93`), i didn't get anything else so far.  
Edit : Wow, i don't know what happened but i got something very weird (several attempts in parallel to try to send the score (or start the round ?) ; i guess an easy fix is to set it on a variable to avoid this case, but that's really weird) :  
```
05:30:54,408 Successfully finished the round and got expected data back:
Scratchpad/16:365:5
05:30:54,409 Level:  12 
EXP:  4140925
Scratchpad/16:366:5
05:30:54,409
Object { response: {…} }
Scratchpad/16:367:5
05:30:55,003 Updating grid
Scratchpad/16:84:4
05:30:55,324 Successfully updated map data on planet: 20
Scratchpad/16:485:4
05:30:55,325 Error getting zone response: 
Object { response: {} }
Scratchpad/16:249:6
05:30:55,327 Waiting 5s and re-sending join attempt(Attempt #0).
Scratchpad/16:84:4
05:30:55,379 Ending Round
Scratchpad/16:84:4
05:30:55,942 Successfully finished the round and got expected data back:
Scratchpad/16:365:5
05:30:55,942 Level:  12 
EXP:  4142125
Scratchpad/16:366:5
05:30:55,943
Object { response: {…} }
Scratchpad/16:367:5
05:30:56,233 Error getting zone response: 
Object { response: {} }
Scratchpad/16:249:6
05:30:56,236 Waiting 5s and re-sending join attempt(Attempt #0).
Scratchpad/16:84:4
05:31:00,782 Error getting zone response: 
Object { response: {} }
Scratchpad/16:249:6
05:31:00,783 Waiting 5s and re-sending join attempt(Attempt #1).
Scratchpad/16:84:4
05:31:01,684 Error getting zone response: 
Object { response: {} }
Scratchpad/16:249:6
05:31:01,685 Waiting 5s and re-sending join attempt(Attempt #1).
```

Harest committed:
Avoiding parallel retries

ensingm2 committed:
Added a bunch of clearTimeouts

ensingm2 committed:
Merge branch 'patch-1' of https://github.com/Harest/saliengame_idler into pr/57

ensingm2 commented:
I more or less did the same timeout changes you did, but there's already a current_timeout variable, and we shouldn't ever be using multiples anyway, so I just used that. Also added the cancelTimeouts and setting current_timeout elsewhere. Needs testing first so not going to merge immediately.

Harest commented:
I don't code often, and i'm not 100% sure how setTimeout works. Afaik it's an id returned that's why i set another variable initially, to avoid losing a timeout we can't cancel anymore. But if i was wrong then that's great ^^.

ensingm2 commented:
Yeah, that's correct thinking. I kept the Interval for the check_game_state its own variable (though I modified it a little to make it cleaner). _**That**_ one should be the only one running concurrently with anything else.

For all other timeouts (exception being ajax error handler and gui initializer), I use the same variable, and clear any pre-existing timeouts beforehand, as there should never be multiples to my knowledge.

I'm just going to let this run overnight before merging, as I'm tired and don't trust myself to not have messed something up massively here.

Harest commented:
Yeah in `checkUnlockGameState()` i used clearInterval() and set the Interval back after the function call because i feared in case the function call is taking too long (> 60s due to network errors), it'd start several loops.  

I'll do the same, just copied the current script we've, we'll see how it went later on.  
Edit : I just did a little edit, removing the clearInterval in `INJECT_leave_planet` to put it in `INJECT_init_battle_selection` instead. It needs to be always up. Right before `check_game_state = setInterval(checkUnlockGameState, 60000);` :  
```
	if (check_game_state !== undefined)
		clearInterval(check_game_state);
```

Harest commented:
I saw several planets switches in the logs (2 or 3), no much errors. It successfully runned 7h in a row so far (from 4M166 to 4M678). I'll apply my little modification quoted above however. So we're sure this little function is always up and running.

Harest committed:
Avoiding checkUnlockGameState() to not be running

With the previous location of clearInterval(check_game_state), the script could potentially be stopped (too many errors when leaving the round for instance). Let's be sure this function is running.

ensingm2 commented:
Why does checkUnnlockGameState need to be running between leaving a planet and re-initializing the battle-selector? That will introduce errors if the Interval runs while you're at the planet selector map.

Also, ran into the undefined 'error' issue in #50 again, which halted my script. I'm going to have to fix that tonight when I get a chance.

Harest commented:
Because of the possibility for leave_round to fail. ~~And technically, i assume if it fails while switching planet, it'll call back switch_planet and it should work without any issue.~~ Nevermind too, we're checking the current state of the game and it works only from the CBattleSelectionState atm. I guess we can just add another condition to `return;` if it's neither an instanceof CBattleSelectionState AND an instanceof CPlanetSelectionState ? Edit : And yes, like you said calling only join_planet in the case where it's an instanceof CPlanetSelectionState already.

I also saw a `window.gGame.m_State.m_PlanetData.zones[zone_id] is undefined`, though it didn't block the script. It comes from wait_for_end that's executed during a planet switch it seems. As the function to check if a better planet exists is executed in start_round.  
~~We should probably `clearTimeout(current_timeout);` in INJECT_leave_round when it succeeded.~~ Nevermind it's already the case, hum...  
```
11:29:19,125 Left Zone #79
Scratchpad/1:85:4
11:29:19,129
TypeError: window.gGame.m_State.m_PlanetData.zones[zone_id] is undefined[En savoir plus]
Scratchpad/1:441:2
11:29:21,638 Initializing Planet Selection Menu.
```

ensingm2 commented:
Ah yeah I guess you're right. Was worried abou tit trying to call SwitchBestZone or INJECT_start_round from the planet selection, which will result in errors, but I don't believe it'll ever reach that point.

It'll still hang though, because if it calls INJECT_switch_planet, it'll immediately return as the first line in that function checks to make sure you're at the battle map (It does this so it doesn't try to leave a planet when none exists). There should probably be a conditional for switch_planet to just run join_planet if it detects it's already at the battle map.

Harest committed:
Minor fix in GetBestPlanet (avoid useless switch)

Missing brackets to register max diff of an active planet

ensingm2 committed:
Allow calls to INJECT_switch_planet from planet menu

Harest commented:
(wait before merging, got a little fix to for the undefined issue)

Harest committed:
Fix undefined case of target_zone

If current_game_id were already defined, we'd set target_zone to -1 in INJECT_leave_round and it'd trigger a potential undefined error.

ensingm2 committed:
Add live-toggle of auto_planet_switch.active

ensingm2 committed:
Merge branch 'patch-1' of https://github.com/Harest/saliengame_idler into pr/57

ensingm2 committed:
Disable auto_switch_planet when anims are disabled

ensingm2 commented:
I feel like despite this still having occasional errors, it's still more stable than master currently? Unless you have a different opinion @Harest, I'll probably let this run an hour or so then merge. Can fix the other issues in other PRs.

Harest committed:
Fix a potential lock in a game area

A rare case where the game server would let you in a game area despite a successful score reported. Added a check on json response headers and leave_round if needed.

Harest commented:
@ensingm2 Well, i runned the script from ~7:00am until i had the error (#62) unrelated to the script itself but from a server-side issue at 2:23am so yeah, i'd say this is quite stable. We've a lot of improvements in there. Until another strange error coming from the server, i think it's a quite good functional script now. We'll just have to adapt for the bosses eventually. If there's a bonus score, etc.

Harest committed:
Fix the other undefined case on target_zone

I could occur during a planet switch with leave_round called before. This one wasn't blocking.
